server/src/main/java/org/elasticsearch/script/field/WriteField.java

```diff
@@ -641,26 +799,55 @@ protected void resolveDepthFlat() {
         int index = path.indexOf('.');
         int lastIndex = 0;
         String segment;
+        if (index == -1 && container instanceof Map<?, ?> mapContainer) {
+            if (mapContainer.containsKey(path)) {
+                onResolutionSuccess(path);
+            } else {
+                onResolutionErrorNotFound(lastIndex);
+            }
+            return;
+        }
 
         while (index != -1) {
             segment = path.substring(lastIndex, index);
-            Object value = container.get(segment);
-            if (value instanceof Map<?, ?> map) {
-                container = (Map<String, Object>) map;
+            Object value = getFromContainer(segment, MISSING);
+            if (value instanceof Map<?, ?> || value instanceof List<?>) {
+                container = value;
                 lastIndex = index + 1;
                 index = path.indexOf('.', lastIndex);
             } else {
                 // Check rest of segments as a single key
                 String rest = path.substring(lastIndex);
-                if (container.containsKey(rest)) {
-                    leaf = rest;
-                } else {
-                    leaf = null;
+                if (container instanceof Map<?, ?> mapContainer && mapContainer.containsKey(rest)) {
+                    onResolutionSuccess(rest);
+                    return;
+                }
+                // Check if last segment is an array index
```

Comment:
If `a.b.c` was nested, this would actually work. But you're right, in this version, it doesn't work if the key is flat. This isn't necessarily a regression as flat keys don't work at all at the moment.

We could change the algorithm back to what it was before https://github.com/elastic/elasticsearch/pull/96786/commits/4a0cd6578cedc7a550f08637f34fdf6d74b5f3c8 to support that case. But that would imply more complexity and more overhead when resolving fields.

See also this discussion for more context: https://github.com/elastic/elasticsearch/pull/96786#discussion_r1231401988